src/range_proof.rs

```diff
@@ -299,46 +302,183 @@ where
         for (i, item) in a_ri.iter_mut().enumerate() {
             *item += d[i] * y_powers[bit_length * aggregation_factor - i] + z;
         }
-        let mut alpha1 = alpha;
         let mut z_even_powers = Scalar::ONE;
         for j in 0..aggregation_factor {
             z_even_powers *= z_square;
-            for (k, alpha1_val) in alpha1.iter_mut().enumerate().take(extension_degree) {
+            for (k, alpha1_val) in alpha.iter_mut().enumerate().take(extension_degree) {
                 *alpha1_val += z_even_powers * witness.openings[j].r[k] * y_powers[bit_length * aggregation_factor + 1];
             }
         }
 
-        // Calculate the inner product
-        transcript.domain_separator(b"Bulletproofs+", b"Inner Product Proof");
-        let mut ip_data = InnerProductRound::init(
-            statement.generators.gi_base_iter().cloned().collect(),
-            statement.generators.hi_base_iter().cloned().collect(),
-            statement.generators.g_bases().to_vec(),
-            statement.generators.h_base().clone(),
-            a_li,
-            a_ri,
-            alpha1,
-            y_powers,
-            &mut transcript,
-            statement.seed_nonce,
-            aggregation_factor,
-        )?;
-        loop {
-            let _result = ip_data.inner_product(rng);
-            if ip_data.is_done() {
-                return Ok(RangeProof {
-                    a: a.compress(),
-                    a1: ip_data.a1_compressed()?,
-                    b: ip_data.b_compressed()?,
-                    r1: ip_data.r1()?,
-                    s1: ip_data.s1()?,
-                    d1: ip_data.d1()?,
-                    li: ip_data.li_compressed()?,
-                    ri: ip_data.ri_compressed()?,
-                    extension_degree: statement.generators.extension_degree(),
-                });
+        let mut gi_base: Vec<P> = statement.generators.gi_base_iter().cloned().collect();
+        let mut hi_base: Vec<P> = statement.generators.hi_base_iter().cloned().collect();
+        let g_base = statement.generators.g_bases();
+        let h_base = statement.generators.h_base();
+
+        let rounds = usize::try_from(full_length.ilog2()).map_err(|_| ProofError::SizeOverflow)?;
+        let mut li = Vec::<P>::with_capacity(rounds);
+        let mut ri = Vec::<P>::with_capacity(rounds);
+
+        let mut n = full_length;
+        let mut round = 0;
+
+        // Perform the inner-product folding rounds
+        while n > 1 {
+            n /= 2;
+
+            // Split the vectors for folding
+            let (a_lo, a_hi) = a_li.split_at(n);
+            let (b_lo, b_hi) = a_ri.split_at(n);
+            let (gi_base_lo, gi_base_hi) = gi_base.split_at(n);
+            let (hi_base_lo, hi_base_hi) = hi_base.split_at(n);
```

Comment:
```suggestion
            // Split the vectors for folding
            fn checked_split_at<T>(my_vec: &[T], n: usize) -> Result<(&[T], &[T]), ProofError> {
                if n <= my_vec.len() {
                    Ok(my_vec.split_at(n))
                } else {
                    Err(ProofError::InvalidLength("Index out of bounds".to_string()))
                }
            }
            let (a_lo, a_hi) = checked_split_at(&a_li, n)?;
            let (b_lo, b_hi) = checked_split_at(&a_ri, n)?;
            let (gi_base_lo, gi_base_hi) = checked_split_at(&gi_base, n)?;
            let (hi_base_lo, hi_base_hi) = checked_split_at(&hi_base, n)?;
```

Reply:
Doesn't this basically treat the split the same as before from a panic perspective? The slice-based version has the same issue.

Reply:
Updated in a new commit.